src/parameter_input.cpp

```diff
@@ -60,72 +60,215 @@
 #include <string>
 #include <vector>
 
+#include <toml.hpp>
+
 #include "globals.hpp"
 #include "utils/error_checking.hpp"
 
 namespace parthenon {
 
-//----------------------------------------------------------------------------------------
-// ParameterInput constructor
+std::string ParameterInput::Path_(const std::string block, const std::string name) {
+  if (!std::count(block.begin(), block.end(), '/')) {
+    if (name == "") {
+      return block;
+    } else {
+      return block + "." + name;
+    }
+  } else {
+    std::string b(block);
+    std::replace(b.begin(), b.end(), '/', '.');
+    if (name == "") {
+      return b;
+    } else {
+      return b + "." + name;
+    }
+  }
+}
+
+void ParameterInput::Merge(toml::table &a, const toml::table &b, bool check_dups) {
+  b.for_each([&](const toml::key &key, auto &&el) {
+    recursive_merge(a, b, key, el, check_dups);
+  });
+}
+
+void ParameterInput::recursive_get_paths(toml::table &a, toml::path prefix,
+                                const toml::key &key, std::vector<toml::path> &paths) {
+  if (a[key].is<toml::table>()) {
+    toml::table &achild = a[key].ref<toml::table>();
+    toml::path block = (prefix != toml::path("")) ? toml::path(prefix.append(key.str()))
+                                                  : toml::path(key.str());
+    achild.for_each([&](const toml::key &key, auto &&el) {
+      recursive_get_paths(achild, block, key, paths);
+    });
+  } else {
+    paths.push_back(toml::path(prefix.append(key.str())));
+  }
+}
+std::vector<std::string> ParameterInput::GetAllPaths(toml::table &a) {
```

Comment:
should there be a version of this that just returns all paths for the internal table?